cluster_pipeline.go

```diff
@@ -0,0 +1,130 @@
+package redis
+
+// ClusterPipeline is not thread-safe.
+type ClusterPipeline struct {
+	commandable
+
+	cmds   []Cmder
+	client *ClusterClient
+	closed bool
+}
+
+// ClusterPipeline creates a new pipeline which is able to execute commands
+// against multiple shards. Be aware that multi-pipelines are
+// not able to follow redirects and you may receive errors if a node
+// fails or if keys are moved before/during pipeline execution.
+func (c *ClusterClient) ClusterPipeline() *ClusterPipeline {
+	pipe := &ClusterPipeline{
+		client: c,
+		cmds:   make([]Cmder, 0, 10),
+	}
+	pipe.commandable.process = pipe.process
+	return pipe
+}
+
+func (c *ClusterPipeline) process(cmd Cmder) {
+	c.cmds = append(c.cmds, cmd)
+}
+
+func (c *ClusterPipeline) Close() error {
+	c.closed = true
+	return nil
+}
+
+func (c *ClusterPipeline) Discard() error {
+	if c.closed {
+		return errClosed
+	}
+	c.cmds = c.cmds[:0]
+	return nil
+}
+
+// Exec always returns list of commands and error of the first failed
+// command if any.
+func (c *ClusterPipeline) Exec() ([]Cmder, error) {
+	if c.closed {
+		return nil, errClosed
+	}
+
+	cmds := c.cmds
+	c.cmds = make([]Cmder, 0, 10)
+	if len(cmds) == 0 {
+		return []Cmder{}, nil
+	}
+
+	pipes := &clusterPipelines{make(map[string]*Pipeline), c.client}
+
+	// Assign commands to pipelines
+	for _, cmd := range cmds {
+		pipes.process(cmd)
+	}
+	pipes.exec()
+
+	// Detect MOVED errors, try again
+	for i := 0; i < c.client.opt.getMaxRedirects(); i++ {
+		for _, cmd := range cmds {
+			if err := cmd.Err(); err != nil {
+				if moved, _, addr := c.client.hasMoved(err); moved {
+					cmd.reset()
+					pipes.processOn(addr, cmd)
+				}
+			}
+		}
+		if pipes.len() == 0 {
+			break
+		}
+		c.client.forceReload()
+		pipes.exec()
+	}
+
+	// Search for the first error
+	for _, cmd := range cmds {
+		if err := cmd.Err(); err != nil {
+			return cmds, err
+		}
+	}
+	return cmds, nil
+}
+
+// ------------------------------------------------------------------------
+
+type clusterPipelines struct {
```

Comment:
I didn't want to assign this to the pipeline directly as this only lives on Exec and is much easier to test on its own.